### PR TITLE
FOGL-1052.fix - replacement PR (for PR#546) for unit tests for services/common/utils.py

### PR DESCRIPTION
**Pull request by amarendra-dianomic**

### CODE REVIEW

**Comment by Singhal-Vaibhav** [comment]
It will be great to have the test results and code coverage in the description

File: python/foglamp/services/common/utils.py [hunk 1]
```diff
@@ -18,26 +18,33 @@
...
+async def ping_service(service, loop=None):
     attempt_count = 1
     """Number of current attempt to ping url"""
 
+    loop = asyncio.get_event_loop() if loop is None else loop
     url_ping = "{}://{}:{}/foglamp/service/ping".format(service._protocol,
                                                         service._address,
                                                         service._management_port)
-    async with aiohttp.ClientSession() as session:
+    # print(service.__str__())
```
Comment (praveen-garg) [inline python/foglamp/services/common/utils.py]: remove this print

File: python/foglamp/services/common/utils.py [hunk 2]
```diff
@@ -18,26 +18,33 @@
...
+
+    async with aiohttp.ClientSession(loop=loop) as session:
         while attempt_count < _MAX_ATTEMPTS + 1:
             try:
+                # print(url_ping)
```
Comment (praveen-garg) [inline python/foglamp/services/common/utils.py]: remove this print

File: python/foglamp/services/common/utils.py [hunk 3]
```diff
@@ -18,26 +18,33 @@
...
                 async with session.get(url_ping) as resp:
                     res = await resp.json()
+                    # print(res)
```
Comment (praveen-garg) [inline python/foglamp/services/common/utils.py]: remove this print

File: python/foglamp/services/common/utils.py [hunk 4]
```diff
@@ -18,26 +18,33 @@
...
                     if res["uptime"] is not None:
                         break
-            except:
+            except Exception as ex:
+                # print(str(ex))
```
Comment (praveen-garg) [inline python/foglamp/services/common/utils.py]: remove this print

File: python/foglamp/services/common/utils.py [hunk 5]
```diff
@@ -46,18 +53,23 @@
...
+async def shutdown_service(service, loop=None):
+    loop = asyncio.get_event_loop() if loop is None else loop
+
     try:
         _logger.info("Shutting down the %s service %s ...", service._type, service._name)
         url_shutdown = "{}://{}:{}/foglamp/service/shutdown".format(service._protocol, service._address,
                                                                     service._management_port)
-        async with aiohttp.ClientSession() as session:
+        async with aiohttp.ClientSession(loop=loop) as session:
+            # print(url_shutdown)
```
Comment (praveen-garg) [inline python/foglamp/services/common/utils.py]: remove this print

File: python/foglamp/services/common/utils.py [hunk 6]
```diff
@@ -46,18 +53,23 @@
...
             async with session.post(url_shutdown) as resp:
                 status_code = resp.status
-                assert 200 == resp.status
-    except (Exception, AssertionError) as ex:
+                text = await resp.text()
+                # print(status_code)
```
Comment (praveen-garg) [inline python/foglamp/services/common/utils.py]: remove this print

File: python/foglamp/services/common/utils.py [hunk 7]
```diff
@@ -46,18 +53,23 @@
...
+                if not status_code == 200:
+                    raise Exception(message=text)
+    except Exception as ex:
+        # print("", str(ex))
```
Comment (praveen-garg) [inline python/foglamp/services/common/utils.py]: remove this print

File: tests/unit/python/foglamp/services/common/test_services_common_utils.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            # WHEN the service is pinged with a valid URL
+            with patch.object(utils._logger, "info") as log:
+                service = ServiceRecord("d", "test", "Southbound", "http", server.host, 1, server.port)
+                url_ping = "{}://{}:{}/foglamp/service/ping".format(service._protocol, service._address, service._management_port)
+                log_params = 'Ping received for Service %s id %s at url %s', service._name, service._id, url_ping
+                resp = await utils.ping_service(service, loop=loop)
+
+        # THEN ping response is received
+        assert resp is True
+        log.assert_called_once_with(*log_params)
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/common/test_services_common_utils.py]: indent above 3 lines

File: tests/unit/python/foglamp/services/common/test_services_common_utils.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            with patch.object(utils._logger, "error") as log:
+                service = ServiceRecord("d", "test", "Southbound", "http", server.host+"1", 1, server.port)
+                url_ping = "{}://{}:{}/foglamp/service/ping".format(service._protocol, service._address, service._management_port)
+                log_params = 'Ping not received for Service %s id %s at url %s attempt_count %s', service._name, service._id, \
+                       url_ping, utils._MAX_ATTEMPTS+1
+                resp = await utils.ping_service(service, loop=loop)
+
+        # THEN ping response is NOT received
+        assert resp is False
+        log.assert_called_once_with(*log_params)
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/common/test_services_common_utils.py]: indent above 3 lines

Comment (praveen-garg) [reply tests/unit/python/foglamp/services/common/test_services_common_utils.py]: log call count should be 2

File: tests/unit/python/foglamp/services/common/test_services_common_utils.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                service = ServiceRecord("d", "test", "Southbound", "http", server.host, 1, server.port)
+                url_shutdown = "{}://{}:{}/foglamp/service/shutdown".format(service._protocol, service._address,
+                                                                            service._management_port)
+                log_params1 = "Shutting down the %s service %s ...", service._type, service._name
+                log_params2 = 'Service %s, id %s at url %s successfully shutdown', service._name, service._id, url_shutdown
+                resp = await utils.shutdown_service(service, loop=loop)
+
+        # THEN shutdown returns success
+        assert resp is True
+        log.assert_called_with(*log_params2)
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/common/test_services_common_utils.py]: indent above 3 lines

File: tests/unit/python/foglamp/services/common/test_services_common_utils.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+                with patch.object(utils._logger, "exception") as log2:
+                    service = ServiceRecord("d", "test", "Southbound", "http", server.host, 1, server.port+1)
+                    log_params1 = "Shutting down the %s service %s ...", service._type, service._name
+                    log_params2 = 'Error in Service shutdown %s, %s', service._name, ' '
+                    resp = await utils.shutdown_service(service, loop=loop)
+
+        # THEN shutdown fails
+        assert resp is False
+        log1.assert_called_with(*log_params1)
+        log2.assert_called()
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/common/test_services_common_utils.py]: indent above 3 lines,
`log1.assert_called_once_with(*log_params1)` and 
`log2.assert_called_once_with(*log_params2)`

**Pull request merged**